bfdriver/efi/efi_vmexits.cpp

```diff
@@ -0,0 +1,382 @@
+
+#include <bfvmm/vcpu/vcpu_factory.h>
+#include <bfvmm/debug/serial/serial_ns16550a.h>
+#include <eapis/vcpu/arch/intel_x64/vcpu.h>
+#include <eapis/hve/arch/intel_x64/ept/memory_map.h>
+#include <eapis/hve/arch/intel_x64/ept/helpers.h>
+
+#include <bfvmm/memory_manager/arch/x64/map_ptr.h>
+
+using namespace eapis::intel_x64;
+
+namespace efi
+{
+
+// -----------------------------------------------------------------------------
+// Handlers
+// -----------------------------------------------------------------------------
+
+static bool
+handle_cpuid(gsl::not_null<bfvmm::intel_x64::vmcs *> vmcs)
+{
+    if (vmcs->save_state()->rax == 0xBF01 || vmcs->save_state()->rax == 0xBF00) {
+        // bfvmm handles
+        return false;
+    }
+
+    auto leaf = vmcs->save_state()->rax;
+    auto ret =
+        ::x64::cpuid::get(
+            gsl::narrow_cast<::x64::cpuid::field_type>(vmcs->save_state()->rax),
+            gsl::narrow_cast<::x64::cpuid::field_type>(vmcs->save_state()->rbx),
+            gsl::narrow_cast<::x64::cpuid::field_type>(vmcs->save_state()->rcx),
+            gsl::narrow_cast<::x64::cpuid::field_type>(vmcs->save_state()->rdx)
+        );
+
+    vmcs->save_state()->rax = ret.rax;
+    vmcs->save_state()->rbx = ret.rbx;
+    vmcs->save_state()->rdx = ret.rdx;
+    if (leaf == 0x00000001) {
+        vmcs->save_state()->rcx = (ret.rcx) & (~(1UL << 26)) & (~(1UL << 27)) & (~(1UL << 5));
+        vmcs->save_state()->rdx = (ret.rdx) & (~(1UL << 12));
+    }
+    else if ((leaf & 0xC0000000) == 0xC0000000) {
+        vmcs->save_state()->rax = 0;
+        vmcs->save_state()->rcx = 0;
+        vmcs->save_state()->rdx = 0;
+    }
+    else if (leaf == 0x0000000A) {
+        vmcs->save_state()->rax = 0;
+        vmcs->save_state()->rcx = 0;
+    }
+    else {
+        vmcs->save_state()->rcx = ret.rcx;
+    }
+
+    return advance(vmcs);
+}
+
+static bool
+handle_rdmsr(gsl::not_null<vmcs_t *> vmcs)
+{
+    auto msr = gsl::narrow_cast<::x64::msrs::field_type>(vmcs->save_state()->rcx);
+
+    switch (msr) {
+        case 0x613:
+        case 0x619:
+            vmcs->save_state()->rax = 0;
+            vmcs->save_state()->rdx = 0;
+            return advance(vmcs);
+    }
+
+    return false;
+}
+
+static bool
+handle_wrmsr(gsl::not_null<vmcs_t *> vmcs)
+{
+    auto msr = gsl::narrow_cast<::x64::msrs::field_type>(vmcs->save_state()->rcx);
+    uint64_t val = ((vmcs->save_state()->rdx) << 0x20) | ((vmcs->save_state()->rax) & 0xFFFFFFFF);
+
+    if (msr == ::intel_x64::msrs::ia32_efer::addr) {
+        if (::intel_x64::vmcs::secondary_processor_based_vm_execution_controls::unrestricted_guest::is_disabled()) {
+            return false;
+        }
+
+        if (val & ::intel_x64::msrs::ia32_efer::lme::mask) {
+            val |= ::intel_x64::msrs::ia32_efer::lma::mask;
+            ::intel_x64::vmcs::guest_cr0::set((0x60000030ULL) | (1ULL << 0) | (1ULL << 31) | (1ULL << 16));
```

Comment:
This needs to use the intrinsics namespace for guest_cr0. Rather than using the raw integers, use
```cpp
auto cr0 = ::intel_x64::vmcs::guest_cr0::get();
cr0 = ::intel_x64::vmcs::guest_cr0::protection_enable::enable(cr0);
cr0 = ::intel_x64::vmcs::guset_cr0::write_protect::enable(cr0);
cr0 = ::intel_x64::vmcs::guset_cr0::paging::enable(cr0);
...
::intel_x64::vmcs::guest_cr0::set(cr0);

This method should be used for all bit manipulations. In general, read the value once, modify the bits as needed, then set the value once. Probably 98% of the names used in the manual map one-to-one to intrinsics namespaces, so you should be able to grep for "namespace <name>" under bfintrinsics to see the different fields available